src/Evolution/Systems/GrMhd/GhValenciaDivClean/BoundaryConditions/ProductOfConditions.hpp

```diff
@@ -126,30 +288,80 @@ struct ProductOfConditionsImpl<
         Tags::detail::TemporaryReference<
             DeduplicatedGridlessTags,
             tmpl::at<gridless_tags_and_types, DeduplicatedGridlessTags>>...,
-        Tags::detail::TemporaryReference<DeduplicatedTempTags>...>
-        shuffle_refs{gridless_variables..., temp_variables...};
-
-    auto gh_string = gh_condition.dg_ghost(
-        gh_variables..., gh_fluxes..., gamma1, gamma2, lapse, shift,
-        inv_spatial_metric, face_mesh_velocity, normal_covector, normal_vector,
-        gh_int_evolved_variables..., gh_int_prim_variables...,
-        tuples::get<Tags::detail::TemporaryReference<GhInteriorTempTags>>(
-            shuffle_refs)...,
-        gh_int_dt_variables..., gh_int_deriv_variables...,
-        tuples::get<Tags::detail::TemporaryReference<
-            GhGridlessTags, tmpl::at<gridless_tags_and_types, GhGridlessTags>>>(
-            shuffle_refs)...);
-    auto valencia_string = valencia_condition.dg_ghost(
-        valencia_variables..., valencia_fluxes..., lapse, shift,
-        inv_spatial_metric, face_mesh_velocity, normal_covector, normal_vector,
-        valencia_int_evolved_variables..., valencia_int_prim_variables...,
-        tuples::get<Tags::detail::TemporaryReference<ValenciaInteriorTempTags>>(
-            shuffle_refs)...,
-        valencia_int_dt_variables..., valencia_int_deriv_variables...,
-        tuples::get<Tags::detail::TemporaryReference<
-            ValenciaGridlessTags,
-            tmpl::at<gridless_tags_and_types, ValenciaGridlessTags>>>(
-            shuffle_refs)...);
+        Tags::detail::TemporaryReference<DeduplicatedTempTags>...,
+        Tags::detail::TemporaryReference<DeduplicatedInteriorEvolvedTags>...>
+        shuffle_refs{gridless_variables..., temp_variables...,
+                     int_evolved_variables...};
+
+    std::optional<std::string> gh_string{};
+    if constexpr (DerivedGhCondition::bc_type ==
+                      evolution::BoundaryConditions::Type::Ghost or
+                  DerivedGhCondition::bc_type ==
+                      evolution::BoundaryConditions::Type::
+                          GhostAndTimeDerivative) {
+      gh_string = gh_condition.dg_ghost(
+          gh_variables..., gh_fluxes..., gamma1, gamma2, lapse, shift,
+          inv_spatial_metric, face_mesh_velocity, normal_covector,
+          normal_vector,
+          tuples::get<Tags::detail::TemporaryReference<GhInteriorEvolvedTags>>(
+              shuffle_refs)...,
+          gh_int_prim_variables...,
+          tuples::get<Tags::detail::TemporaryReference<GhInteriorTempTags>>(
+              shuffle_refs)...,
+          gh_int_dt_variables..., gh_int_deriv_variables...,
+          tuples::get<Tags::detail::TemporaryReference<
+              GhGridlessTags,
+              tmpl::at<gridless_tags_and_types, GhGridlessTags>>>(
+              shuffle_refs)...);
+    } else {
+      GhDoNothingGhostCondition{}.dg_ghost(
```

Comment:
So the trick here is that one of the use-cases we want to support is when the Valencia system uses only `Ghost` and the GH system uses only `TimeDerivative` -- the cleanest way to support that I could find is to have the product system then have a BC type `GhostAndTimeDerivative`, but then we have the problem where the GH part of the product system needs to provide ghost data (because the boundary correction will be applied regardless), but its corresponding boundary condition is of type `TimeDerivative`.

Does that make sense, or is there a cleaner way you can see of doing this?

Reply:
Oh, yea... I misread the `if` check. Specifically, I missed that it's only checking the GH condition. Thanks!